peps/pep-0748.rst

```diff
@@ -0,0 +1,735 @@
+
+.. code:: pycon
+
+  >>> annotations.call_evaluate_function(field.evaluate_type, annotations.Format.SOURCE)
+  'undefined'
+  >>> annotations.call_evaluate_function(field.evaluate_type, annotations.Format.FORWARDREF)
+  ForwardRef('undefined')
+
+Other variations are possible. For example, we could leave the ``type`` attribute unchanged,
+and only add the ``evaluate_type`` method. This avoids unpleasant surprises where accessing
+``.type`` may throw an exception.
+
+Acknowledgments
+===============
+
+First of all, I thank Larry Hastings for writing :pep:`649`. This PEP modifies some of his
+initial decisions, but the overall design is still his.
+
+I thank Carl Meyer and Alex Waygood for feedback on early drafts of this PEP.
+
+Appendix
+========
+
+Which expressions can be stringified?
+-------------------------------------
+
+:pep:`649` acknowledges that the stringifier cannot handle all expressions. Now that we
+have a draft implementation, we can be more precise about the expressions that can and
+cannot be handled. Below is a list of all expressions in the Python AST that can and
+cannot be recovered by the stringifier. The full list should probably not be added to
+the documentation, but creating it is a useful exercise.
+
+First, the stringifier of course cannot recover any information that is not present in
+the compiled code, including comments, whitespace, parenthesization, and operations that
+get simplified by the AST optimizer.
+
+Second, the stringifier can intercept almost all operations that involve names looked
+up in some scope, but it cannot interject operations that operate fully on constants.
+As a corollary, this also means it is not safe to request SOURCE format on untrusted
+code: Python is powerful enough that it is possible to achieve arbitrary code execution
+even with no access to any globals or builtins. For example:
+
+```
+>>> def f(x: (1).__class__.__base__.__subclasses__()[-1].__init__.__builtins__["print"]("Hello world")): pass
+... 
+>>> annotations.get_annotations(f, format=annotations.Format.SOURCE)
+Hello world
+{'x': 'None'}
+```
+
+The following are supported (sometimes with caveats):
+
+* ``BinOp``
+* ``UnaryOp``
+  * ``Invert`` (``~``), ``UAdd`` (``+``), and ``USub`` (``-``) are supported
+  * ``Not`` (``not``) is not supported
+* ``Dict`` (except when using ``**`` unpacking)
+* ``Set``
+* ``Compare``
+  * ``Eq`` and ``NotEq`` are supported
+  * ``Lt``, ``LtE``, ``Gt``, and ``GtE`` are supported, but the operand may be flipped
+  * ``Is``, ``IsNot``, ``In``, and ``NotIn`` are not supported
+* ``Call`` (except when using ``**`` unpacking)
+* ``Constant`` (though not the exact representation of the constant; for example, escape
+  sequences in strings are lost; hexadecimal numbers are converted to decimal)
+* ``Attribute`` (assuming the value is not a constant)
+* ``Subscript`` (assuming the value is not a constant)
+* ``Starred`` (``*`` unpacking)
+* ``Name``
+* ``List``
+* ``Tuple``
+* ``Slice``
```

Comment:
The sub-lists haven't rendered properly, maybe they need extra indents:

<img width="778" alt="image" src="https://github.com/python/peps/assets/1324225/8960377c-d00d-417f-9e5b-471581150f9a">